Plastic Surgurry

Vimas21 opened the pull request:
:cl: Viim
rcsadd: Adds plastic surgery! Targeting the head while the skull is unbroken, using a scalpel, then hemostat, allows you to change the name and voice of your target. This can trick people and basic huds, but it won't hold up to forensic or body scanner scrutiny. Perform the surgery again to reset to normal!
/:cl:

Spookerton commented:
🤔

PigeonLord commented:
Scalpel, then hemostat is how you clamp bleeders during a surgery, though. I think the steps should be changed, since this would make clamping bleeders when performing head surgery impossible

Vimas21 commented:
> Scalpel, then hemostat is how you clamp bleeders during a surgery, though. I think the steps should be changed, since this would make clamping bleeders when performing head surgery impossible

As it stands, you have to begin surgery first- so the total list is as follows:
Target face + scalpel + hemostat + clamp, and THEN scalpel + hemostat. Similar to fixing a scarred face, where you have to open the face and then use the hemostat.

SierraKomodo commented on code/modules/mob/mob_defines.dm:
```diff
@@ -82,6 +82,8 @@
 	var/next_move = null
 	var/hand = null
 	var/real_name = null
+	var/fake_name = null
+	var/name_faked = FALSE
```
This can probably be shortened to a single var - If `fake_name` is not null, it can be safely assumed the name is faked, thus removing the need for a separate boolean var.

SierraKomodo commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -44,7 +44,10 @@
 		CRASH("Gender datum was null; key was '[(skipjumpsuit && skipface) ? PLURAL : gender]'")
 
-	msg += "<EM>[src.name]</EM>"
+	if(src.name_faked)
+		msg += "<EM>[src.fake_name]</EM>"
+	else
+		msg += "<EM>[src.name]</EM>"
```
Unless there's a locally scoped var defined in the proc of the same name, there's no need for a `src.` prefix in var references.
```suggestion
	if(name_faked)
		msg += "<EM>[fake_name]</EM>"
	else
		msg += "<EM>[name]</EM>"
```

SierraKomodo commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -265,7 +268,10 @@
+		if(src.name_faked)
+			msg += "<span class='danger'>[src.fake_name] [T.has] \a [implant.name] sticking out of [T.his] flesh!</span>\n"
+		else
+			msg += "<span class='danger'>[src] [T.has] \a [implant.name] sticking out of [T.his] flesh!</span>\n"
```
Also, `\The` and span defines.

```suggestion
		if(name_faked)
			msg += SPAN_DANGER("\The [fake_name] [T.has] \a [implant.name] sticking out of [T.his] flesh!") + "\n"
		else
			msg += SPAN_DANGER("\The [src] [T.has] \a [implant.name] sticking out of [T.his] flesh!") + "\n"
```

SierraKomodo commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -277,7 +283,10 @@
 			perpname = id.registered_name
 		else
-			perpname = src.name
+			if(src.name_faked)
+				perpname=src.fake_name
+			else
+				perpname=src.name
```
```suggestion
			if(name_faked)
				perpname=fake_name
			else
				perpname=name
```

SierraKomodo commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -295,7 +304,10 @@
 			perpname = id.registered_name
 		else
-			perpname = src.name
+			if(src.name_faked)
+				perpname=src.fake_name
+			else
+				perpname=src.name
```
```suggestion
			if(name_faked)
				perpname=fake_name
			else
				perpname=name
```

SierraKomodo commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -340,7 +352,10 @@
 	set category = "IC"
 
-	pose =  sanitize(input(usr, "This is [src]. [get_visible_gender() == MALE ? "He" : get_visible_gender() == FEMALE ? "She" : "They"]...", "Pose", null)  as text)
+	if(src.name_faked)
+		pose =  sanitize(input(usr, "This is [src.fake_name]. [get_visible_gender() == MALE ? "He" : get_visible_gender() == FEMALE ? "She" : "They"]...", "Pose", null)  as text)
```
```suggestion
	if(name_faked)
		pose =  sanitize(input(usr, "This is [fake_name]. [get_visible_gender() == MALE ? "He" : get_visible_gender() == FEMALE ? "She" : "They"]...", "Pose", null)  as text)
```

SierraKomodo commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -299,11 +299,13 @@
 /mob/living/carbon/human/proc/get_face_name()
 	var/obj/item/organ/external/H = get_organ(BP_HEAD)
-	if(!H || (H.status & ORGAN_DISFIGURED) || H.is_stump() || !real_name || (MUTATION_HUSK in mutations) || (wear_mask && (wear_mask.flags_inv&HIDEFACE)) || (head && (head.flags_inv&HIDEFACE)))	//Face is unrecognizeable, use ID if able
+	if(!H || name_faked || (H.status & ORGAN_DISFIGURED) || H.is_stump() || !real_name || (MUTATION_HUSK in mutations) || (wear_mask && (wear_mask.flags_inv&HIDEFACE)) || (head && (head.flags_inv&HIDEFACE)))	//Face is unrecognizeable, use ID if able
 		if(istype(wear_mask) && wear_mask.visible_name)
 			return wear_mask.visible_name
 		else if(istype(wearing_rig) && wearing_rig.visible_name)
 			return wearing_rig.visible_name
+		else if(name_faked)
+			return fake_name
 		else
 			return "Unknown"
 	return real_name
```
This will result in cases that should display as `Unknown` for various reasons instead displaying the fake name. This would be better implemented as a separate check when you'd otherwise use the real name.
```suggestion
	if(!H || (H.status & ORGAN_DISFIGURED) || H.is_stump() || !real_name || (MUTATION_HUSK in mutations) || (wear_mask && (wear_mask.flags_inv&HIDEFACE)) || (head && (head.flags_inv&HIDEFACE)))	//Face is unrecognizeable, use ID if able
		if(istype(wear_mask) && wear_mask.visible_name)
			return wear_mask.visible_name
		else if(istype(wearing_rig) && wearing_rig.visible_name)
			return wearing_rig.visible_name
		else
			return "Unknown"
	return name_faked ? fake_name : real_name
```

SierraKomodo commented on code/modules/mob/living/carbon/human/say.dm:
```diff
@@ -124,7 +124,8 @@
 			if(changer && changer.active && changer.voice)
 				voice_sub = changer.voice
-
+	if(name_faked)
+		return fake_name
```
As above - This will cause a fake name to override other mechanics. Attach it to where `real_name` is returned instead.

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+	shock_level = 20
+
+/decl/surgery_step/plastic_surgery/prepare_face/begin_step(mob/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	user.visible_message("[user] starts peeling back the skin around \the [target]'s face with \the [tool].", \
+	"You start peeling back the skin around \the [target]'s face with \the [tool].")
```
- Span defines
- Use `\The` before object references
- `\` for line breaks isn't necessary in this case
- Better formatting of multi-line proc calls

```suggestion
	user.visible_message(
		SPAN_NOTICE("\The [user] starts peeling back the skin around \the [target]'s face with \the [tool]."),
		SPAN_NOTICE("You start peeling back the skin around \the [target]'s face with \the [tool].")
	)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+
+/decl/surgery_step/plastic_surgery/prepare_face/end_step(mob/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	user.visible_message("[user] finishes peeling back the skin around \the [target]'s face with \the [tool].", \
+	"You finish peeling back the skin around \the [target]'s face with \the [tool].")
```
```suggestion
	user.visible_message(
		SPAN_NOTICE("\The [user] finishes peeling back the skin around \the [target]'s face with \the [tool]."),
		SPAN_NOTICE("You finish peeling back the skin around \the [target]'s face with \the [tool].")
	)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+
+/decl/surgery_step/plastic_surgery/prepare_face/fail_step(mob/living/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	var/obj/item/organ/external/affected = target.get_organ(target_zone)
+	user.visible_message("<span class='warning'>[user]'s hand slips, tearing skin on [target]'s face with \the [tool]!</span>", \
+	"<span class='warning'>Your hand slips, tearing skin on [target]'s face with \the [tool]!</span>")
```
```suggestion
	user.visible_message(
		SPAN_WARNING("\The [user]'s hand slips, tearing skin on \the [target]'s face with \the [tool]!"),
		SPAN_WARNING("Your hand slips, tearing skin on \the [target]'s face with \the [tool]!")
	)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+
+/decl/surgery_step/plastic_surgery/reform_face/begin_step(mob/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	user.visible_message("[user] starts molding \the [target]'s face with \the [tool].", \
+	"You starts molding \the [target]'s face with \the [tool].")
```
```suggestion
	user.visible_message(
		SPAN_NOTICE("\The [user] starts molding \the [target]'s face with \the [tool]."),
		SPAN_NOTICE("You start molding \the [target]'s face with \the [tool].")
	)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+
+/decl/surgery_step/plastic_surgery/reform_face/end_step(mob/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	user.visible_message("[user] finishes molding \the [target]'s face with \the [tool].", \
+	"You finish molding \the [target]'s face with \the [tool].")
```
```suggestion
	user.visible_message(
		SPAN_NOTICE("\The [user] finishes molding \the [target]'s face with \the [tool]."),
		SPAN_NOTICE("You finish molding \the [target]'s face with \the [tool].")
	)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+	"You finish molding \the [target]'s face with \the [tool].")
+	if(!target.name_faked)
+		var/new_name = sanitizeSafe(input(user, "Select a new name for \the [target].") as text|null, MAX_NAME_LEN)
+		user.visible_message("[user] molds \the [target]'s face into the spitting image of [new_name]!", \
+		"You mold \the [target]'s face into the spitting image of [new_name]!")
```
```suggestion
		user.visible_message(
			SPAN_NOTICE("\The [user] molds \the [target]'s face into the spitting image of [new_name]!"),
			SPAN_NOTICE("You mold \the [target]'s face into the spitting image of [new_name]!")
		)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+	user.visible_message("[user] finishes molding \the [target]'s face with \the [tool].", \
+	"You finish molding \the [target]'s face with \the [tool].")
+	if(!target.name_faked)
+		var/new_name = sanitizeSafe(input(user, "Select a new name for \the [target].") as text|null, MAX_NAME_LEN)
```
This needs additional validation after receiving the response from the input box to ensure the surgery step is still valid - Don't want someone filling this prompt in 3 minutes later or after being moved out of range and suddenly someone's name changes, for example.

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+		target.name_faked=FALSE
+		target.fake_name=null
+		user.visible_message("[user] returns \the [target]'s face back to normal!", \
+		"You return \the [target]'s face back to normal!")
```
```suggestion
		user.visible_message(
			SPAN_NOTICE("\The [user] returns \the [target]'s face back to normal!"),
			SPAN_NOTICE("You return \the [target]'s face back to normal!")
		)
```

SierraKomodo commented on code/modules/surgery/face.dm:
```diff
@@ -29,11 +29,102 @@
+/decl/surgery_step/plastic_surgery/reform_face/fail_step(mob/living/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	var/obj/item/organ/external/affected = target.get_organ(target_zone)
+	user.visible_message("<span class='warning'>[user]'s hand slips, tearing skin on [target]'s face with \the [tool]!</span>", \
+	"<span class='warning'>Your hand slips, tearing skin on [target]'s face with \the [tool]!</span>")
```
```suggestion
	user.visible_message(
		SPAN_WARNING("\The [user]'s hand slips, tearing skin on \the [target]'s face with \the [tool]!"),
		SPAN_WARNING("Your hand slips, tearing skin on \the [target]'s face with \the [tool]!")
	)
```

SierraKomodo commented on code/modules/mob/living/carbon/hallucinations.dm:
```diff
@@ -158,10 +158,17 @@
 				phrases += list("What did you come here for[add]?","Don't touch me[add].","You're not getting out of here[add].", "You are a failure, [pick(names)].","Just kill yourself already, [pick(names)]")
 			message = pick(phrases)
-			to_chat(holder,"<span class='game say'><span class='name'>[talker.name]</span> [holder.say_quote(message)], <span class='message'><span class='body'>\"[message]\"</span></span></span>")
+			if(talker.fake_name)
+				to_chat(holder,"<span class='game say'><span class='name'>[talker.fake_name]</span> [holder.say_quote(message)], <span class='message'><span class='body'>\"[message]\"</span></span></span>")
+			else
+				to_chat(holder,"<span class='game say'><span class='name'>[talker.real_name]</span> [holder.say_quote(message)], <span class='message'><span class='body'>\"[message]\"</span></span></span>")
```
May be worth adding a `display_name` var that's set to `fake_name` or `real_name` as necessary, then dropping that var into the `to_chat()` call to reduce on the copy+paste, i.e. 
```dm
var/display_name = talker.fake_name ? talker.fake_name : talker.real_name
```

SierraKomodo commented on code/modules/mob/living/carbon/hallucinations.dm:
```diff
@@ -158,10 +158,17 @@
 		else
-			to_chat(holder,"<B>[talker.name]</B> points at [holder.name]")
-			to_chat(holder,"<span class='game say'><span class='name'>[talker.name]</span> says something softly.</span>")
+			if(talker.fake_name)
+				to_chat(holder,"<B>[talker.fake_name]</B> points at [holder.name]")
+				to_chat(holder,"<span class='game say'><span class='name'>[talker.fake_name]</span> says something softly.</span>")
+			else
+				to_chat(holder,"<B>[talker.name]</B> points at [holder.name]")
+				to_chat(holder,"<span class='game say'><span class='name'>[talker.name]</span> says something softly.</span>")
```
Same as above, though note the existing code here uses `name` instead of `real_name`.

SierraKomodo reviewed (dismissed):
Also still items from the previous review.

Vimas21 committed:
Plastic Surgurry


Further Work

Vocal Chords are done. All the examine shit is going to be HARD.
All Done! ^^

Charlie's Inferno
Fix For Siesie


Apply suggestions from code review

Co-authored-by: Sierra Brown <SierraKomodo@users.noreply.github.com>
e


d


f

SierraKomodo reviewed (commented):
Works in testing, checking with Spook on final say since he's expressed a dislike of the concept.

Wiki update's ready here as well: https://wiki.baystation12.net/User:Sierrakomodo/Surgery-PR-30776#Plastic_Surgery

Spookerton reviewed (approved):
Hate, but probably won't see use outside raider/merc. So we'll see how that goes, I guess.

SierraKomodo commented:
Wiki updated - https://wiki.baystation12.net/Surgery#Plastic_Surgery
